### DISCUSSION TITLE
How to access the flutter ModalRoute<T>?

### DISCUSSION THREAD

**Discussion by Adam-Langley** [Milad-Akarie/auto_route_library#1927] [General]
I have an application consisting of nested routers, and when state changes in my root route, I need to force an update in a child route.
Usually (when not using auto_route) I would access the ModalRoute, and call `changedExternalState` - triggering a rebuild.
How can I access the native ModalRoute object constructed within auto_route, so that I can call this method?

Thanks!

**Comment by Milad-Akarie** [comment]
Hey @Adam-Langley unless the route you're working with is a tab-page you should be able to access it the same? 
how do you usually access the ModalRoute?

**Comment by Adam-Langley** [comment]
Hey @Milad-Akarie - thanks for responding.

Usually, I am just pushing auto routes...

```
 AutoRouter.of(context)
    .navigate(MyAutoRoute());
```

However, the "MyAutoRoute" is an instance of PageRouteInfo, not a ModalRoute... I dont see how to get access to the internal native class that would normally expose the 'changedExternalState' function.

I have some dialogs that dont fit the auto_route way of doing things, so I wrap them in a ModalRoute - and that *does* expose the `changedExternalState` - so I have no trouble sending them message to ask them to rebuild.

How do I do it with auto_route?

I realise I could inject some argument exposing some kind of function pointer I can use to trigger setState... however I'd rather use what's already in the framework...

**Reply by Milad-Akarie** [reply]
@Adam-Langley so your goal is to rebuild a modal from outside it's scope.

Guess you can try to re-navigate to the same page if its the top page of course

**Reply by Adam-Langley** [reply]
Hey - yes indeed that's the goal. Ok - so mainly I was checking that I wasn't completely blind, and not seeing the function declaration....

**Reply by Milad-Akarie** [reply]
@Adam-Langley `changedExternalState` is a protected method and I wouldn't recommend using it, idealy I'd do the refresh task using a state managment solution or some event bus. that being said you can access the top route inside the navigator doing something like this
router.pagelessRoutesObserver.current?

**Reply by Adam-Langley** [reply]
Thanks @Milad-Akarie - I'll look into that - it might do the trick
